### PR TITLE
[to_review] implementation of DASVM

### PR DESCRIPTION
**Pull request by BuenoRuben**
I am still ensure about how to implement the class for dasvm

### CODE REVIEW

**Comment by BuenoRuben** [comment]
for now, I am just trying to create something that can be used in da_pipelines

**Comment by rflamary** [comment]
You shoudl call teh class `DASVMEstimator`this is not an adapter.

**Comment by BuenoRuben** [comment]
oh, yes, I am correcting this right now

File: skada/_dasvm.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+import math
+
+from skada.base import BaseEstimator
+from skada.utils import check_X_y_domain, source_target_split
+from sklearn.base import clone
+
+from sklearn.svm import SVC
+
+
+class DASVMEstimator(BaseEstimator):
```
Comment (tgnassou) [inline skada/_dasvm.py:14]: You have to describe better the docstings. Link to the paper, parameters ... Look at other methods to see what's you have to do

File: skada/_dasvm.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    """
+    Simple dasvm estimator
+    """
+
+    __metadata_request__fit = {'sample_domain': True}
+    __metadata_request__transform = {'sample_domain': True, 'allow_source': True}
+
+    def __init__(
+            self, base_estimator=None,
```
Comment (tgnassou) [inline skada/_dasvm.py:23]: To be more consistent with the rest of the library, can you put one parameters per line

File: skada/_dasvm.py (hunk 1 above)
Comment (tgnassou) [inline skada/_dasvm.py:14]: Maybe we can wait to have the PR of Rémi to have the BaseDAEstimator? wdyt @kachayev @rflamary

File: skada/_dasvm.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            Returns self.
+        """
+        X, y, sample_domain = check_X_y_domain(X, y, sample_domain)
+        Xs, Xt, ys, _ = source_target_split(
+            X, y, sample_domain=sample_domain
+        )
+
+        n = Xs.shape[0]
+        m = Xt.shape[0]
+        self.c = np.unique(ys).shape[0]  # number of classes
```
Comment (tgnassou) [inline skada/_dasvm.py:128]: put more explicit name as c -> n_classes

File: skada/_dasvm.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+        # doing the selection on the labeled data
+        self._find_points_next_step(
+            index_source_deleted, decisions_s)
+        # doing the selection on the semi-labeled data
+        self._find_points_next_step(
+            index_target_added, decisions_ta)
+
+        i = 0
+        while (
```
Comment (tgnassou) [inline skada/_dasvm.py:167]: Try to avoid while loop. Use instead for loop with a stopping criteria

File: skada/_dasvm.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+            # We finally only need to change the list
+            for l in I:
+                # indices_list[l] is False at that point
+                indices_list[l] = True
+
+    def _get_X_y(
+            self, new_estimator, index_target_added, index_source_deleted, Xs, Xt, ys
+            ):
+        # Allow to get the X and y arrays at a state of the algorithm
```
Comment (tgnassou) [inline skada/_dasvm.py:72]: Put it in docstring of the function

File: skada/_dasvm.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            for j in range(len(index_target_added)):
+                if index_target_added[j]:
+                    x = Xt[j]
+                    if new_estimator.predict(
+                            [x]) != old_estimator.predict([x]):
+                        # index_target_added[j] should be True
+                        index_target_added[j] = False
+
+            # look at those that have not been discarded
+            decisions_s = self.get_decision(new_estimator, Xs, index_source_deleted)
```
Comment (tgnassou) [inline skada/_dasvm.py:187]: decision_source decision_target or decision_s decision_t, but not decision_ta

File: skada/tests/test_dasvm.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+import pytest
+
+
+@pytest.mark.parametrize(
+    "label, n, m",
+    [("binary", 7, 5), ("binary", 6, 6), ("binary", 5, 7), ("multiclass", 7, 5)],
+)
+def test_dasvm_estimator(label, n, m):
+    """
```
Comment (tgnassou) [inline skada/tests/test_dasvm.py:18]: don't need docstring

File: skada/_dasvm.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    __metadata_request__transform = {'sample_domain': True, 'allow_source': True}
+
+    def __init__(
+            self, base_estimator=None,
+            k=3, Stop=1_000, **kwargs
+            ):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
```
Comment (tgnassou) [inline skada/_dasvm.py:29]: The parameter should be in the class docstring not in the __init__ docstring

File: skada/_dasvm.py (hunk 1 above)
Comment (BuenoRuben) [reply skada/_dasvm.py:14]: yes, I have done it, I just haven't pushed it yet, waiting for more important changes

File: examples/methods/plot_dasvm.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+axis[0].set_title("source data points")
+
+axis[1].scatter(Xt[:, 0], Xt[:, 1], c=yt)
+axis[1].set_xlim(xlim)
+axis[1].set_ylim(ylim)
+axis[1].set_title("target data points")
+
+figure.suptitle("data points", fontsize=20)
+
+E = DASVMEstimator(
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:61]: estimator

File: skada/_dasvm.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+import math
+
+from skada.base import BaseEstimator
+from skada.utils import check_X_y_domain, source_target_split
+from sklearn.base import clone
+
+from sklearn.svm import SVC
+
+
+class DASVMEstimator(BaseEstimator):
```
Comment (tgnassou) [inline skada/_dasvm.py:25]: Use the `baseDAestimator`

Comment (tgnassou) [reply skada/_dasvm.py:25]: Like that you do'nt have to put the metadata request

File: skada/_dasvm.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        """
+
+    __metadata_request__fit = {'sample_domain': True}
+    __metadata_request__transform = {'sample_domain': True, 'allow_source': True}
+
+    def __init__(
+            self,
+            base_estimator=None,
+            k=3,
+            Stop=1_000,
```
Comment (tgnassou) [inline skada/_dasvm.py:48]: check if there is a similar parameters in the toolbox but maybe `max_iter`. Try to not use capital letter

File: skada/_dasvm.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+        # This is the list of the indices from the
+        # points from Xs that have been discarded
+        index_source_deleted = np.array([False]*n)
+        # This is the list of the indices from the
+        # points from Xt that have been added
+        index_target_added = np.array([False]*m)
+
+        # In the first step, the SVC is fitted on the source
+        X = Xs
```
Comment (tgnassou) [inline skada/_dasvm.py:156]: Maybe call them X_train, y_train not to get confused with the input ?

File: skada/tests/test_dasvm.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        n_samples_target=m,
+        shift="covariate_shift",
+        noise=None,
+        label=label,
+    )
+    X, y, sample_domain = check_X_y_domain(X, y, sample_domain)
+    Xs, Xt, ys, yt = source_target_split(
+        X, y, sample_domain=sample_domain)
+
+    clf_dsvm = DASVMEstimator(k=5).fit(X, y, sample_domain=sample_domain)
```
Comment (tgnassou) [inline skada/tests/test_dasvm.py:29]: clf_dasvm

File: skada/__init__.py [hunk 14]
```diff
@@ -38,6 +38,7 @@
     TransferComponentAnalysisAdapter,
     TransferComponentAnalysis,
 )
+from ._dasvm import DASVMEstimator
```
Comment (tgnassou) [inline skada/__init__.py:41]: call the submodule `_self_labeling.py`

File: skada/_dasvm.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+
+The dsvm method comes from [1].
+
+.. [1]  Domain Adaptation Problems: A DASVM Classification
+        Technique and a Circular Validation Strategy
+        Lorenzo Bruzzone, Fellow, IEEE, and Mattia Marconcini, Member, IEEE
+
+"""
+# Author: Ruben Bueno <ruben.bueno@polytechnique.edu>
+# dasvm implementation
```
Comment (tgnassou) [inline skada/_dasvm.py:13]: add license BSD 3-Clause

File: skada/_dasvm.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+        """
+        X = np.concatenate((Xs[~index_source_deleted], Xt[index_target_added]))
+        semi_labels = new_estimator.predict(
+            Xt[index_target_added])
+        y = np.concatenate((ys[~index_source_deleted], semi_labels))
+        return X, y
+
+    def _get_decision(self, new_estimator, X, indices_list):
+        """
+        We look at the points that have either not been discarded or not been added
```
Comment (tgnassou) [inline skada/_dasvm.py:103]: make the docstring a bit shorter

**Review by tgnassou — CHANGES_REQUESTED** [review]

File: examples/methods/plot_dasvm.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
+"""
+Plot for the dasvm estimator
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:2]: Change title of example for DASVM classifier example

File: examples/methods/plot_dasvm.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+======================
+
+This example illustrates the dsvm method from [21].
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:5]: dsvm -> DASVM

File: examples/methods/plot_dasvm.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+# however the estimator has been created only for SVC
+base_estimator = SVC()
+
+target_marker = "s"
+source_marker = "o"
+xlim = (-1.5, 2.4)
+ylim = (-1, 1.3)
+
+# %%
+# We generate our dataset
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:43]: We generate 2D dataset with 2 classes

File: examples/methods/plot_dasvm.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+# ------------------------------------------
+#
+# We generate a simple 2D covariate shift dataset.
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:46]: We generate a simple 2D  dataset from moon distributionwhere source and target are not taken from the same location in the moons. Something like that

File: examples/methods/plot_dasvm.py [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+
+
+# %%
+#     Plots of the dataset
+# ------------------------------------------
+#
+# As we can see, the source and target datasets have different
+# distributions for the points' positions but have the same
+# labels for the same x-values.
+# We are then in the case of covariate shift
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:69]: .

File: examples/methods/plot_dasvm.py [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+
+# %%
+#     Usage of the DASVMClassifier
+# ------------------------------------------
+#
+# The main problem here is that we only know the distribution of the points
+# from the target dataset, our goal is to label it.
+#
+# The algorithm of the dasvm consist in fitting multiple base_estimator (SVC) by:
+#     - removing from the training dataset (if possible)
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:94]: The itemize is weird on the example

File: examples/methods/plot_dasvm.py [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+axis[-1].legend(
+    handles=[margin_line, decision_boundary])
+
+# %%
+#     Labeling the target dataset
+# ------------------------------------------
+#
+# Here we show 4 states from our algorithm, At first we are only given source
+# data points with label (which are circle, in colors showing the label), and
+# taget datapoints that have no labels (which are represented as squares, in
```
Comment (tgnassou) [inline examples/methods/plot_dasvm.py:201]: target

File: skada/_self_labeling.py [hunk 24]
```diff
@@ -0,0 +1,268 @@
+"""
+    dasvm estimator
```
Comment (tgnassou) [inline skada/_self_labeling.py:2]: DASVM Classifier

File: examples/methods/plot_dasvm.py (hunk 22 above)
Comment (BuenoRuben) [reply examples/methods/plot_dasvm.py:94]: yes..., I solved it at the same moment you  sent me the review, it should be fixed ^^

**Review by tgnassou — APPROVED** [review]

**Pull request merged**